af_ktls.c

```diff
@@ -358,30 +379,569 @@ static void tls_update_senpage_ctx(struct tls_sock *tsk, size_t size)
+
+				rxm->accum_len += uneaten;
+				eaten += uneaten;
+				WARN_ON(eaten != orig_len);
+				break;
+			} else if (len < 0) {
+				/* Data does not appear to be a TLS record
+				 * Make userspace handle it
+				 */
+				goto decryption_fail;
+			} else if (len <= (ssize_t)head->len -
+					  skb->len - rxm->offset) {
+				/* Length must be into new skb (and also
+				 * greater than zero)
+				 */
+				goto decryption_fail;
+			}
 
+			rxm->full_len = len;
+		}
+
+		extra = (ssize_t)(rxm->accum_len + uneaten) - rxm->full_len;
+
+		if (extra < 0) {
+			/* Message not complete yet. */
+			if (rxm->full_len - rxm->accum_len >
+			    tcp_inq((struct sock *)tsk)) {
+				/* Don't have the whole messages in the socket
+				 * buffer. Set tsk->rx_need_bytes to wait for
+				 * the rest of the message. Also, set "early
+				 * eaten" since we've already buffered the skb
+				 * but don't consume yet per tcp_read_sock.
+				 * If function returns 0, does not consume
+				 */
+
+				/* Wait. Why doesn't this code path just set
+				 * eaten? Then tcp_read_sock will eat and
+				 * profit!
+				 */
+				if (!rxm->accum_len) {
+					/* Start RX timer for new message */
+					/*kcm_start_rx_timer(tsk); */
+				}
+
+				tsk->rx_need_bytes = rxm->full_len -
+						       rxm->accum_len;
+				rxm->accum_len += uneaten;
+				rxm->early_eaten = uneaten;
+				desc->count = 0; /* Stop reading socket */
+				break;
+			}
+			rxm->accum_len += uneaten;
+			eaten += uneaten;
+			WARN_ON(eaten != orig_len);
+			break;
+		}
+
+		/* Positive extra indicates ore bytes than needed for the
+		 * message
+		 */
+
+		WARN_ON(extra > uneaten);
+
+
+		ret = decrypt_skb(tsk, head);
+
+		/* Hurray, we have a new message! */
+		tsk->rx_skb_head = NULL;
+		eaten += (uneaten - extra);
+
+		sock_queue_rcv_skb(sk, head);
+	}
+done:
+	if (cloned_orig)
+		kfree_skb(orig_skb);
+	return eaten;
+decryption_fail:
+	kfree_skb(skb);
+	desc->error = -EBADMSG;
+	tsk->rx_skb_head = NULL;
+	desc->count = 0;
+	goto done;
+}
+
+static int tls_tcp_read_sock(struct tls_sock *tsk)
+{
+	read_descriptor_t desc;
+
+	desc.arg.data = tsk;
+	desc.error = 0;
+	desc.count = 1; /* give more than one skb per call */
+
+	/* sk should be locked here, so okay to do tcp_read_sock */
+	tcp_read_sock(tsk->socket->sk, &desc, tls_tcp_recv);
+
+	return desc.error;
+}
+
+static void do_tls_data_ready(struct tls_sock *tsk)
+{
+	int ret;
+
+	ret = tls_tcp_read_sock(tsk);
+	if (ret == -ENOMEM) /* No memory. Do it later */
+		queue_work(tls_wq, &tsk->recv_work);
+	/*queue_delayed_work(tls_wq, &tsk->recv_work, 0); */
+	/* TLS couldn't handle this message. Pass it directly to userspace */
+	else if (ret == -EBADMSG)
+		tls_err_abort(tsk);
+
+}
+
+/* Called with lower socket held */
 static void tls_data_ready(struct sock *sk)
 {
 	struct tls_sock *tsk;
 
-	xprintk("--> %s", __FUNCTION__);
+	xprintk("--> %s", __func__);
 
 	read_lock_bh(&sk->sk_callback_lock);
-
+	//TODO: forgot to lock tsk?
```

Comment:
Actually, recvmsg() does not acquire the underlying socket lock anymore. But we still cannot lock_sock(tsk) because tls_data_ready is called in softirq context. While the SKB's are protected by sk_buff_head locks, I think another lock is needed to protect sk->sk_err. Hence why I think we still need to lock_sock(). I think that tls_data_ready should check if the socket is locked (sock_owned_by_user), and if so, defer the work to a workqueue.